daemon/ZilliqaDaemon.cpp

```diff
@@ -15,95 +15,157 @@
+  if (m_pids[name].empty()) {
+    m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Looking for new "
+          << name << " process..." << endl;
+    vector<pid_t> tmp = ZilliqaDaemon::GetProcIdByName(name);
+
+    for (const pid_t& i : tmp) {
+      m_died[i] = false;
+      m_pids[name].push_back(i);
+      m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+            << "Started monitoring new process " << name << " with PiD: " << i
+            << endl;
     }
+
+    return;
   }
 
-  string lastLine = "";
-  getline(logFile, lastLine);
+  for (const pid_t& pid : m_pids[name]) {
+    // If sig is 0 (the null signal), error checking is performed but no signal
+    // is actually sent
+    int w = kill(pid, 0);
 
-  // log<<"Line :"<<lastLine<<endl;
+    if (w < 0) {
+      if (errno == EPERM) {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+              << "Daemon does not have permission "
+              << "Name: " << name << " Id: " << pid << endl;
+      } else if (errno == ESRCH) {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Process died "
+              << "Name: " << name << " Id: " << pid << endl;
+        m_died[pid] = true;
+      } else {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+              << "Kill failed due to " << errno << " Name: " << name
+              << "Id: " << pid << endl;
+      }
+    }
 
-  logFile.close();
+    if (m_died[pid]) {
+      auto it = find(m_pids[name].begin(), m_pids[name].end(), pid);
 
-  return lastLine;
+      if (it != m_pids[name].end()) {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str() << "Not monitoring "
+              << pid << " of " << name << endl;
+        m_pids[name].erase(it);
+      }
+
+      StartNewProcess();
+      m_died.erase(pid);
+    }
+  }
 }
 
-void setIsDs(const pid_t& pid, ofstream& log) {
-  string s = ReadLastLine(Path[pid], log);
+string ZilliqaDaemon::CurrentTimeStamp() {
+  auto tm = chrono::system_clock::to_time_t(chrono::system_clock::now());
+  char* t = ctime(&tm);
 
-  if (s == " DS BACKUP NOW") {
-    isDS[pid] = true;
-  } else {
-    isDS[pid] = false;
+  if (t[strlen(t) - 1] == '\n') {
+    t[strlen(t) - 1] = '\0';
   }
+
+  return "[" + string(t) + "] : ";
 }
 
-vector<pid_t> getProcIdByName(const string& procName, ofstream& log) {
+string ZilliqaDaemon::Execute(const string& cmd) {
+  array<char, 128> buffer{};
+  string result;
+  shared_ptr<FILE> pipe(popen(cmd.c_str(), "r"), pclose);
+
+  if (!pipe) throw std::runtime_error("popen() failed!");
+
+  while (!feof(pipe.get())) {
+    if (fgets(buffer.data(), 128, pipe.get()) != nullptr)
+      result += buffer.data();
+  }
+
+  return result;
+}
+
+bool ZilliqaDaemon::DownloadPersistenceFromS3() {
+  string output;
+  m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+        << "downloading persistence from S3" << endl;
+  output = Execute("python " + m_curPath + download_incr_DB_script);
```

Comment:
Can we update the script `download_incr_DB.py`  to support python3 interpreter?

Reply:
After some checking, `python3` cannot be used to run `download_incr_DB.py` because of the lack of `request` package. I'll leave it to later's refactor. @sandipbhoir